cmd/argocd/commands/root.go

```diff
@@ -27,6 +29,52 @@ func initConfig() {
 	cli.SetLogLevel(cmdutil.LogLevel)
 }
 
+func NewDefaultArgoCDCommand() *cobra.Command {
+	return NewDefaultArgoCDCommandWithArgs(ArgoCDCLIOptions{
+		PluginHandler: NewDefaultPluginHandler([]string{"argocd"}),
+		Arguments:     os.Args,
+	})
+}
+
+func NewDefaultArgoCDCommandWithArgs(o ArgoCDCLIOptions) *cobra.Command {
+	cmd := NewCommand()
+
+	if o.PluginHandler == nil {
+		return cmd
+	}
+
+	// the first argument will be the binary, followed by other arguments
+	if len(o.Arguments) > 1 {
+		cmdPathPieces := o.Arguments[1:]
+
+		// Try to find a valid Argo CD command that matches the arguments provided (e.g., foo in the case of argocd foo)
+		// If it finds a command, it continues without invoking the plugin.
+		// If it doesn't find the command (err is non-nil), it means foo isn't a built-in Argo CD command,
+		// so it might be a plugin like argocd-foo.
+		if _, _, err := cmd.Find(cmdPathPieces); err != nil {
```

Comment:
This will introduce an overhead in all argocd commands. To avoid this problem, I would suggest a modification in the proposal to only execute plugin commands in a dedicated subcommand. 
For example, instead of running plugins as:
`argocd some-plugin arg1 arg2`
declare a dedicated subcommand to invoke plugins as:
`argocd plugin some-plugin arg1 arg2`

cc @christianh814

Reply:
Curious to know is it like a really big deal if all the argocd commands have to go through this? IMO, It doesn't look like a big burden but I'm definitely open to opinions.

Reply:
Two main reasons why I am suggesting this change are:
- Will greatly simplify the code. This is a big deal as it reduces the possible edge cases and increases maintainability.
- Avoid unnecessary loop that all `argocd` commands would have to execute. Some companies are very sensitive by the duration `argocd` commands take to execute. While saving a few milliseconds in a single execution looks irrelevant, it can become a significant issue when executed millions of times.